### PR TITLE
🐛  Hotifx sample swap issues

### PR DESCRIPTION
**Pull request by dmiller15**
## Description

As defined https://github.com/d3b-center/bixu-tracker/issues/2049 we're experiencing sample swapping when using the old_tumor/normal_name inputs. These changes make the renaming process error proof as we're now matching the new samples to the old ones before swapping.

Closes https://github.com/d3b-center/bixu-tracker/issues/2066
Closes https://github.com/d3b-center/bixu-tracker/issues/2077

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Previously erroneous run now works: https://cavatica.sbgenomics.com/u/kfdrc-harmonization/kf-reference-pipeline/tasks/39ffe06d-de74-4ea2-97b9-25f89a8d3848/#

Compare the results in above against the pre-swap intermediates found here: https://cavatica.sbgenomics.com/u/kfdrc-harmonization/kf-reference-pipeline/files/#q?page=1&label_id=swap

```
### MANTA 

BEFORE
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  1224650_D1      1182220_D1
AFTER
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  BS_24RR82Q8     BS_PD37THK0

### Strelka2

BEFORE
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  NORMAL  TUMOR
AFTER
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  BS_24RR82Q8     BS_PD37THK0

### Mutect2

BEFORE
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  1182220_D1      1224650_D1
AFTER
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  BS_PD37THK0     BS_24RR82Q8


### Lancet

BEFORE
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  1182220_D1      1224650_D1
AFTER
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  BS_PD37THK0     BS_24RR82Q8

### Vardict

BEFORE
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  1182220_D1      1224650_D1
AFTER
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  BS_PD37THK0     BS_24RR82Q8
```

- [x] Sample swap workflow: https://cavatica.sbgenomics.com/u/kfdrc-harmonization/kf-reference-pipeline/tasks/364e6526-9b87-452a-b30a-a901baa9799e/

**Test Configuration**:
* Environment:
* Test files:

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [x] I have committed any related changes to the PR

### CODE REVIEW

File: sub_workflows/kfdrc_manta_sub_wf.cwl [hunk 1]
```diff
@@ -63,18 +47,35 @@ steps:
     out: [output_sv, small_indels]
 
   rename_manta_samples:
-    run: ../tools/bcftools_reheader_vcf.cwl
+    run: ../tools/bcftools_reheader_samples_index.cwl
+    when: $(inputs.old_tumor_name != null && inputs.old_tumor_name != null)
```
Comment (migbro) [inline sub_workflows/kfdrc_manta_sub_wf.cwl:51]: These say the same thing - maybe you meant and normal?

File: sub_workflows/kfdrc_vardict_sub_wf.cwl [hunk 2]
```diff
@@ -96,13 +96,20 @@ steps:
     out: [merged_vcf]
 
   rename_vcf_samples:
-    when: $(inputs.old_tumor_name != null)
-    run: ../tools/bcftools_reheader_vcf.cwl
+    run: ../tools/bcftools_reheader_samples_index.cwl
+    when: $(inputs.old_tumor_name != null && inputs.old_tumor_name != null)
```
Comment (migbro) [inline sub_workflows/kfdrc_vardict_sub_wf.cwl:100]: Same comment as for manta sub

File: workflow/kfdrc_production_lancet_wf.cwl [hunk 3]
```diff
@@ -44,6 +44,9 @@ inputs:
   output_basename: { type: 'string', doc: "String value to use as basename for outputs" }
   wgs_or_wxs: { type: { type: enum, name: sex, symbols: ["WGS", "WXS"] }, doc: "Select if this run is WGS or WXS" }
 
+  old_tumor_name: { type: 'string?', doc: "If `SM:` sample name in te align file is different than `input_tumor_name`, you **must** provide it here"}
```
Comment (migbro) [inline workflow/kfdrc_production_lancet_wf.cwl:47]: typo - `te` should be `the`

File: workflow/kfdrc_production_manta_wf.cwl [hunk 4]
```diff
@@ -43,6 +43,9 @@ inputs:
   hg38_strelka_tbi: { type: 'File?', doc: "Tabix index for hg38_strelka_bed" }
   output_basename: { type: string, doc: "String value to use as basename for outputs" }
 
+  old_tumor_name: { type: 'string?', doc: "If `SM:` sample name in te align file is different than `input_tumor_name`, you **must** provide it here"}
```
Comment (migbro) [inline workflow/kfdrc_production_manta_wf.cwl:46]: same typo

**Review by migbro — CHANGES_REQUESTED** [review]
Some typos and a possible bug in logic. Still not sure though how the changes catch the issue in which the sort step might reorder the samples...

File: workflow/kfdrc_production_lancet_wf.cwl (hunk 3 above)
Comment (dmiller15) [reply workflow/kfdrc_production_lancet_wf.cwl]: Got it

File: workflow/kfdrc_production_manta_wf.cwl (hunk 4 above)
Comment (dmiller15) [reply workflow/kfdrc_production_manta_wf.cwl]: It was everywhere!

File: sub_workflows/kfdrc_vardict_sub_wf.cwl (hunk 2 above)
Comment (dmiller15) [reply sub_workflows/kfdrc_vardict_sub_wf.cwl]: Got it.

**Comment by dmiller15** [comment]
@migbro The way this works is that the new tool uses the secondary methodology for `bcftools reheader --sample`: https://samtools.github.io/bcftools/bcftools.html#reheader

> new sample names, one name per line, in the same order as they appear in the VCF file. Alternatively, only samples which need to be renamed can be listed as "old_name new_name\n" pairs separated by whitespaces, each on a separate line.

So the old implementation is the one where we only provide the `new_name` and treat the first sample as NORMAL in all cases. This fix uses the alternative approach where we declare BOTH the `new_name` AND the `old_name`. In the original implementation, `bcftools` would always replace the first sample in the VCF with the first sample in the `sample_list.txt` file. In this implementation, `bcftools` only replace `old_name` with `new_name`. If it cannot find `old_name` it will not put in the `new_name`. Additionally, the order of the `sample_list.txt` file no longer matters. All lines in the file can replace any sample in the VCF.

**Review by migbro — APPROVED** [review]
Nice work! I wonder if that matched old/new sample list functionality was always there...much safer!

**Review by sickler-alex — APPROVED** [review]

**Review by migbro — APPROVED** [review]
I think it looks good. We could try a more "surgical" approach, but it'd likely end up being more complicated and probably not worth saving maybe a few hundred bucks over the lost time and effort, which is a greater cost.

**Pull request merged**